### PR TITLE
My hw 03

### PR DESCRIPTION
**Pull request by DmitryS1989**

### CODE REVIEW

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class Homework3
+  def task2(logs)
+    result = []
```
Comment (alexshgov) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb:5]: avoid using the temp variable `result`

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    logs.split("\n").each do |string|
+      result << make_output(string) if check_format?(string)
+    end
+    result
+  end
+
+  private
+
+  def check_format?(string)
+    string.match?(%r{^(?:[0-9]{1,3}\.){3}[0-9]{1,3} - - \[\d*/\w*/\d*:\d*:\d*:\d* \+\d*\] "\w* /\w*/\d/\w*})
```
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb:15]: let's move regex to constants

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb (hunk 1 above)
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_02.rb:5]: Split your log here, and do all the necessary work by parsing in an iterator over the split data

```suggestion
    lines = logs.split
    
    lines.select { |line| what you want to select }. map |line| do
      selected lines manipulation 
    end
```

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_03.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require 'time'
+class Homework3
+  def task3(logs)
```
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_03.rb:5]: I will give you advice on how to create a good solution, so try refactoring your code:
Split log
Select the necessary rows that match the condition
Go through them with an iterator and select the necessary data (I advise you to use a regular expression)
Make a difference calculation method (Time.parse is handy to use)
create a condition if the selected data exists, then pass it to the created method for calculating the difference in events

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class Homework3
+  def task4(string, result = {})
```
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb:4]:
```suggestion
  def task4(string)
```

try not to change it

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+    result[:letters] = string.downcase.count('a-z')
```
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb:5]:
```suggestion
    letters = string.count('a-zA-Z')
```

File: HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    result[:digits] = string.downcase.count('0-9')
+    result
```
Comment (B0RlS) [inline HW03/Dmitry_Susha/dmitry_susha_hw_03_t_04.rb:7]:
```suggestion
    { letters: letters, digits: digits }
```

**Pull request merged**